### PR TITLE
[otbn,rtl,dv] Replace zeroing in secure wipe with writing random data

### PR DESCRIPTION
**Pull request by andreaskurth**
As reasoned in https://github.com/lowRISC/opentitan/issues/13499, the second round of a secure wipe should not
write zeroes but PRNG-provided values.  This should make FI attacks that
rely on a known (or even zero) register state much more difficult.
Before that second round using PRNG-provided values, the PRNG has to be
reseeded.  This is needed so the PRNG state that provided the values for
the first wipe cannot be inferred from reading out registers after the
wipe.

The main commit in this PR (the last one) adds an URND reseed between
the two rounds of a secure wipe and changes the second round from zeroing
to using URND-provided values, like the first round.  This is mainly implemented
in `otbn_start_stop_control`, which controls the secure wipe.
Additionally, the bignum ALU and MAC are slightly modified as follows:
- In `otbn_alu_bignum`, the `sec_wipe_zero_i` input is no longer used to
  zero the `MOD` register.  It is still used to zero the flags, though.
- In `otbn_mac_bignum`, the `sec_wipe_zero_i` input has been removed.
  The MAC only used that input to zero the `ACC` register, which no
  longer happens.

On OTBN DV, the main commit makes the following changes:
- Extend `otbn_core_model`, which models the URND request, to perform
  the URND reseed between the two rounds of a secure wipe.  This
  significantly increases the complexity of that part of the model, but
  the behavior is still coded independently from RTL (except the number
  of cycles for one wipe round, which is dictated by the RTL).
- Change `otbn_scoreboard` to wait for an expected alert for up to 150
  cycles, as the URND reseed in the middle of a secure wipe potentially
  delays the raising of an alert.
- Modify the registers in `otbnsim` to be *invalid* after a secure wipe.
  This is required because the registers are no longer zero after the
  secure wipe, yet the simulator does not know the random values written
  during secure wipe.
- Modify and extend the `_step_wiping()` function of `otbnsim` to model
  two rounds of secure wipe with an URND refresh between them.

## To be discussed

OTBN's simulator does not model the random values that are written during secure wipe. (Before this change, this was not an issue because each secure wipe ended with zeros in all registers.) This creates two practical problems described below.

A solution to both problems could be to model the random values in the simulator precisely as in RTL. This would further increase the complexity of the simulator and reduce its abstraction level, though, so I would refrain from doing this unless we consider this to be the only viable solution.

### Checks after secure wipe

One check to ensure the secure wipe works as intended was to compare the registers of the simulator and the RTL for equality after a secure wipe. This no longer works with this change, because after a secure wipe RTL registers contain random values that the simulator does not know. *Some* checks on the outcome of a secure wipe are required, though.

This PR:
- removes the equality comparison of the registers *after* the secure wipe. (The registers are still compared after *program execution* and thus *before* the secure wipe.)
- modifies the comparison of the register write traces: A value of `x` now compares equal to any value. This is used to compare the traces produced during secure wipe, where the *set* of written registers must be equal between RTL and simulator but the *value* written to registers that are wiped non-zero is ignored.

### Instructions that write only part of a register

Some instructions, such as `bn.mulqacc.so` (are there more?), write only one half of a register. When such an instruction writes a register that has not been set to a value in a way that is independent of the initial value of any register that is wiped non-zero, the outcome will differ between simulator and RTL. This is usually a transient issue, because a program will eventually write the entire register. However, it would cause a mismatch in the traces.

This PR initializes the destination WDRs of the `bn.mulqacc.so` instructions in `smoke_test`, so that this problem does not surface in the smoke test.

In the long run, we may want to model partially unknown registers in the simulator, so that a simulator register could be partially unknown, appear as such in the trace, and compare equal to the RTL register provided the known part matches.

### CODE REVIEW

File: hw/ip/otbn/rtl/otbn_start_stop_control.sv
Comment (vogelpi) [inline hw/ip/otbn/rtl/otbn_start_stop_control.sv:156]: Same here: use `begin ... end`

File: hw/ip/otbn/rtl/otbn_mac_bignum.sv
Comment (vogelpi) [inline hw/ip/otbn/rtl/otbn_mac_bignum.sv:196]: Nit: also the comment can be removed.

**Review by vogelpi — COMMENTED** [review]
Thanks for working on this @andreaskurth . The RTL changes look good to me, I just have some nits. Maybe the `wipe_after_urnd_refresh_q` should be multi bit to make skipping the first/second wipe more difficult.

Your suggestions for the DV challenges sound reasonable to me but I think also others should chime in @GregAC , @ctopal , @prajwalaputtappa .

**Comment by GregAC** [comment]
I agree, modeling the random values written for the secure wipe isn't worth it. Can I just confirm that there is a check to ensure the second write from the secure wipe gives a value that differs from the first write?

Could we add a check to ensure we're not outputting Xs from register writes from the RTL during the secure wipe? Currently our checking looks like it'd pass if there was some bug writing Xs to the registers during the secure wipes.

> Some instructions, such as bn.mulqacc.so (are there more?), write only one half of a register

Only `bn.mulqacc.so` does this

**Review by vogelpi — APPROVED** [review]
From the RTL side, this looks good to me. There is one minor thing that can be done in a follow-up. It shouldn't block this PR.

Regardings Greg's questions:
> 1. Can I just confirm that there is a check to ensure the second write from the secure wipe gives a value that differs from the first write?

Yes, this is the case (see also [this commit](https://github.com/lowRISC/opentitan/pull/13724/commits/905d10caf7549f9dfc5f7272c6d3bfed584452a9))
If EDN provides the same two values for both URND seeds, the smoke test now failes with:
```
Writing simulation traces to sim.fst
ERROR: Mismatch between RTL and ISS trace entries: Repeated identical RTL lines for key `ACC'.
  RTL entry is:
    V 
    > ACC: 0xe5ae8a2c_2b495801_bf29a059_326cc666_4d0e6b05_0a515f14_1d8d4e43_7ead90ae
    > ACC: 0xe5ae8a2c_2b495801_bf29a059_326cc666_4d0e6b05_0a515f14_1d8d4e43_7ead90ae
```

> 2. Could we add a check to ensure we're not outputting Xs from register writes from the RTL during the secure wipe?

This sounds reasonable. Should we do it as a follow-up? I could give it a try this week. This would allow us to merge this fairly complicated PR as is (Andreas is away this week) and unblock a couple of other PRs. WDYT @GregAC ?

I think it would be good if @ctopal could review this from a DV angle please.

**Comment by GregAC** [comment]
> This sounds reasonable. Should we do it as a follow-up? I could give it a try this week. This would allow us to merge this fairly complicated PR as is (Andreas is away this week) and unblock a couple of other PRs. WDYT @GregAC ?

Sounds good to me

**Review by GregAC — APPROVED** [review]

File: hw/ip/otbn/dv/model/otbn_core_model.sv
Comment (ctopal) [inline hw/ip/otbn/dv/model/otbn_core_model.sv:206]: Just for clarification, this assumes in the case of reset of model, we are going to start the whole process in which we do two wipe related URND requests and one URND request after we start execution, right?

**Comment by ctopal** [comment]
>>Can I just confirm that there is a check to ensure the second write from the secure wipe gives a value that differs from the first write?
>
>If EDN provides the same two values for both URND seeds, the smoke test now failes with:
>```
>Writing simulation traces to sim.fst
>ERROR: Mismatch between RTL and ISS trace entries: Repeated identical RTL lines for key `ACC'.
>  RTL entry is:
>    V 
>    > ACC: 0xe5ae8a2c_2b495801_bf29a059_326cc666_4d0e6b05_0a515f14_1d8d4e43_7ead90ae
>    > ACC: 0xe5ae8a2c_2b495801_bf29a059_326cc666_4d0e6b05_0a515f14_1d8d4e43_7ead90ae
>```

I thought the question was more about what does OTBN do if we receive two identical URND seeds regardless of how our trace checker throws an error about it, or maybe I got it wrong. As far as I know URND doesn't share the countermeasure of RND that throws an error in the case of two identical seeds. So, how does RTL understand in the case of identical URND seeds between secure wipe round one and round two?

**Review by ctopal — APPROVED** [review]
LGTM on the DV side with one question for clarification. Thanks a lot @andreaskurth for clearly explaining the whole structure, I really appreciate it!

**Comment by vogelpi** [comment]
> > > Can I just confirm that there is a check to ensure the second write from the secure wipe gives a value that differs from the first write?
> > 
> > 
> > If EDN provides the same two values for both URND seeds, the smoke test now failes with:
> > ```
> > Writing simulation traces to sim.fst
> > ERROR: Mismatch between RTL and ISS trace entries: Repeated identical RTL lines for key `ACC'.
> >  RTL entry is:
> >    V 
> >    > ACC: 0xe5ae8a2c_2b495801_bf29a059_326cc666_4d0e6b05_0a515f14_1d8d4e43_7ead90ae
> >    > ACC: 0xe5ae8a2c_2b495801_bf29a059_326cc666_4d0e6b05_0a515f14_1d8d4e43_7ead90ae
> > ```
> 
> I thought the question was more about what does OTBN do if we receive two identical URND seeds regardless of how our trace checker throws an error about it, or maybe I got it wrong. As far as I know URND doesn't share the countermeasure of RND that throws an error in the case of two identical seeds. So, how does RTL understand in the case of identical URND seeds between secure wipe round one and round two?

Ah, I see. Two identical URND seeds or RND words (both 256 bits wide) aren't detected in hardware. But we have checks at lower granularity further upstream: For RND, there are 32-bit checks inside OTBN and 128-bit checks in EDN (plus more checks upstream). For URND, we have the 128-bit checks inside EDN and more checks upstream inside CSRNG. In addition there are mechanisms to detect spurious ACKs from EDN. This should make it pretty hard for someone to force two 256-bit seeds being equal.

Also, it's worth noting that XoShiRo PRNG used for URND interpretes the 256 input bits as 4 64-bit words and then combines these words together (using XOr, SHIft and ROtations). This results in pretty good diffusion. As a result, unless all 256 bits of two subsequent seeds are equal, the PRNG will quickly produce different outputs.

File: hw/ip/otbn/dv/model/otbn_core_model.sv
Comment (vogelpi) [reply hw/ip/otbn/dv/model/otbn_core_model.sv:206]: This is my assumption, too. But I leave this for @andreaskurth to answer once he returns.

**Comment by vogelpi** [comment]
Thanks for your reviews @ctopal and @GregAC. And thanks @andreaskurth for the hard work. I am merging this now.

File: hw/ip/otbn/rtl/otbn_start_stop_control.sv
Comment (vogelpi) [reply hw/ip/otbn/rtl/otbn_start_stop_control.sv:233]: This is now done, see #13697.

File: hw/ip/otbn/dv/model/otbn_core_model.sv
Comment (andreaskurth) [reply hw/ip/otbn/dv/model/otbn_core_model.sv:206]: Correct

**Pull request merged**
